pkg/ebpf/events_enrich.go

```diff
@@ -8,172 +8,157 @@ import (
+					for evt := range queues[cgroupId] {
+						if evt.ContainerImage == "" {
+							enrichEvent(evt, i.result)
 						}
-
-						enrichEvent(event, info.Container)
-
-						close(queue)
-						queuesMutex.Lock()
-						delete(staleQueues, cgroupId)
-						queuesMutex.Unlock()
-						out <- event
+						out <- evt
 					}
+					bLock.Lock()
+					delete(enrichDone, cgroupId)
+					delete(enrichInfo, cgroupId)
+					delete(queues, cgroupId)
+					bLock.Unlock()
+					continue
 				}
+				// make sure a queue channel exists for this cgroupId
+				if _, ok := queues[cgroupId]; !ok {
+					queueSize := contQueueSize
+					bLock.Lock()
+					queues[cgroupId] = make(chan *trace.Event, queueSize)
+					enrichDone[cgroupId] = &uberAtom.Bool{}
+					bLock.Unlock()
+					go func(cgroupId uint64) {
+						metadata, err := t.containers.EnrichCgroupInfo(cgroupId)
+						bLock.Lock()
+						enrichInfo[cgroupId] = &enrichResult{metadata, err}
+						enrichDone[cgroupId].Toggle()
+						bLock.Unlock()
+					}(cgroupId)
+				}
+				// enqueue the event and schedule the operation
+				queues[cgroupId] <- event
+				queueReady <- cgroupId
+			case <-ctx.Done():
+				done <- struct{}{}
+				return
 			}
 		}
 	}()
 
+	// queues map reader
 	go func() {
-		for enrich := range enriches {
-			cgroupId := enrich.cgroupId
-			mutex := &sync.Mutex{}
-			mutex.Lock() // place already acquired mutex in the attempted map
-			attemptedMutex.Lock()
-			attempted[cgroupId] = mutex // now attempted[cgroupId] is a transaction, only happens once and has begun
-			attemptedMutex.Unlock()
-			queuesMutex.RLock()
-			queue := queues[cgroupId]
-			queuesMutex.RUnlock()
-			if enrich.err != nil {
-				//only send error if it's not a non-existing cgroup error
-				if enrich.err.Error() != "no cgroup to enrich" {
-					t.handleError(enrich.err)
-				}
-
-				for evt := range queue {
-					select {
-					case out <- evt:
-					case <-done:
-						return
-					}
-
-					//at this point no new events should enter this queue, since the attempt was marked
-					//we need this break condition, otherwise we will wait for new events forever
-					if len(queue) < 1 {
-						break
-					}
-				}
-			} else {
-				//go through the queue and inject the enrichment data that was missing during decoding
-				for evt := range queue {
-					if evt.ContainerID != "" {
-						enrichEvent(evt, enrich.result)
-					}
-					select {
-					case out <- evt:
-					case <-done:
-						return
-					}
-
-					//at this point no new events should enter this queue, since the attempt was marked
-					//we need this break condition, otherwise we will wait for new events forever
-					if len(queue) < 1 {
-						break
+		for { // de-queue events
+			select {
+			case cgroupId := <-queueReady: // queue for received cgroupId is ready
+				bLock.RLock()
+				if !enrichDone[cgroupId].Load() {
+					bLock.RUnlock()
+					// re-schedule the operation if queue is not enriched
+					queueReady <- cgroupId
```

Comment:
So if I understand correctly, this will busy loop until container is enriched because it is wrapped by the above:
`case cgroupId := <-queueReady:`
If that's the case, is it really what we want (busy-looping)?

Reply:
The other option is doing what I previously did which was queueing on enrichments, and then emptying the queues, but that requires coordinating stale queues etc. which is what we wanted to avoid.

Reply:
Plus (to what Nadav said) we're not busy looping during enrichment, we're busy looping in a "simple scheduler" iteration. And it mixes events from multiple containers, if they're coming through channel, giving chance for all containers events to be processed.